### ISSUE TITLE
[t400/t500] Dock is broken

### ISSUE THREAD

**Issue by ghost** [librecore-org/librecore#15]
On the Lenovo T400 and T500, the Advanced Mini Dock hardly works. Only charging and VGA / DVI work, but USB and RS232 do nothing. I haven't tested ethernet yet, but it probably doesn't work either.

The eject button also doesn't work, nor does the LED indicator.

**Comment by ArthurHeymans** [comment]
T400 (T500/W500) code in coreboot is a simple clone of X200, including X200 dock code. Since the T400 is compatible with T60 dock (same dock?), it is probably possible to reuse a large part of that code (both SMI and ACPI). I suspect its a matter of setting right GPIOs and also setup SIO (which is trivial since T60 code is most certainly compatible).

**Comment by ghost** [comment]
Yes, same dock. X200 docks work mostly ok, but are completely different. Only the eject button on the X200 dock doesn't work for some reason, while Fn + F9 does.

**Comment by ghost** [comment]
After testing https://review.coreboot.org/#/c/18054/ I can conclude that:

(Mostly) working:

- USB now works in Linux, but not in GRUB;
- Both VGA and DVI work in Linux as a second screen, but LVDS is disabled in GRUB when VGA is connected (on the X200 the screens are mirrored);
- DVI does not work at all in GRUB, because of #13;
- Audio out works;
- Audio in works (shows up as dock microphone even);
- Ethernet works;
- Serial port + debugging works;
- The power button on the dock works;
- Suspend / resume when docked works.

Not working:

- The undock button on the dock and Fn + F9 do not work;
- The undock led indicators do not work;
- Undocking and then redocking results in a hard lockup.

```
ACPI: \_SB.DOCK: undocking
ACPI: \_SB.DOCK: No _EJ0 support for device
```

Untested:

- S-Video;
- Coaxial SPDIF;
- Parallel port;
- 56k modem.

**Comment by al12gamer** [comment]
Will this work with the T400 and Advanced Dock 2503 model?
